src/pytorch_lightning/strategies/fully_sharded_native.py

```diff
@@ -215,9 +240,21 @@ def setup(self, trainer: "pl.Trainer") -> None:
         assert self.lightning_module is not None
         self.lightning_module._device = self.root_device
 
+        assert isinstance(self.model, pl.LightningModule)
+        self.model = _LightningModuleWrapperBase(self.model)
```

Comment:
This is because @rohitgr7 added the FSDP wrapper around the LM. When we call model.forward() below, we still need to redirect to training_step or whatever *_step internally, based on the stage, and this is what this `_LightningModuleWrapperBase` does.

Reply:
Yes, but why do it different now than before?

Reply:
I think because before, we didn't wrap the lightning module at all. Now that we do, we need to do the same as in DDP.

Reply:
so the input cast in case of precision happens inside the Fullyshardedwrapper, which internally calls the base modules forward method. I did this to ensure we have a forward method available that gets the casted input and then the lightningmodule methods are called accordingly. Previously I guess it wasn't done because the wrap was done by the user manually, so there's always a wrapper around each individual layer.

I think it might work without `_LightningModuleWrapperBase`, but just wanted to be consistent here and avoid any edge cases with casting.

Reply:
That would explain why we wrap with `_LightningModuleWrapperBase`, but why not use `self.lightning_module`?

Reply:
because if we do that, then each layer will act as root w.r.t FSDP and will create this issue:
https://github.com/pytorch/pytorch/blob/67ece03c8cd632cce9523cd96efde6f2d1cc8121/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1464-L1475

also, the tests fail if I use `lightning_module` and use the `*_step` methods directly. The inputs should be passed from the wrapped module first to ensure each sub-layer should not act as root.